<h1 style='font-family: Verdana'>Assignment 1</h1>

There are eight questions in total. Each question is worth 5 marks. The maximum possible score for this assignment is 40 marks. 

Please replace ```### Your code here ###``` or ```# replace the pass with your code``` with the *your solution code*, but **do not remove the existing code or change the names of each function, class, parameters, attributes and methods that are already provided in the question**.

For each question, there will be two open test-cases, and three hidden test-cases. Rounding errors of 1e-05 will be tolerated for `float` number output, ie. if the exact answer is 0.5, your answer will be accepted as long as it is between 0.49999 and 0.50001.

[Rule 1] Please attempt the questions in good faith. If you were to hard-code the answers (for example using a dictionary to map the known inputs and outputs), no credit will be awarded for that.

[Rule 2] Please ensure that you *return* (NOT *print*) the output.

[Rule 3] You are not allowed to import any library. It is okay to use whatever built-in Python functions.

<h2 style='font-family: Verdana'>Question 1 - V fan</h2>

Write a function `vfan`, which takes in a single character, a *positive odd-valued* integer and a second single character. You can assume the two single characters are ASCII directly interpretable by Python.

The function will return a multi-line string of v-shaped fan pattern. The number of lines depends on the input integer.

For example, the call of `vfan('!', 5, '%')` will return: `'!        %\n!!      %%\n!!!    %%%\n!!!!  %%%%\n!!!!!%%%%%'`. 

If the returned string is displayed with the `print` function, it looks like this:
```python
!        %
!!      %%
!!!    %%%
!!!!  %%%%
!!!!!%%%%%
```

In [8]:
def vfan(c1, odd_value, c2) -> str:
    # replace pass with your code
    lines = []
    for i in range(odd_value):
        line = c1 * (i + 1) + ' ' * (2 * (odd_value - i - 1)) + c2 * (i + 1)
        lines.append(line)
    return '\n'.join(lines)



In [9]:
# Sample 1.1
vfan('!', 5, '%')

'!        %\n!!      %%\n!!!    %%%\n!!!!  %%%%\n!!!!!%%%%%'

In [10]:
# the following code is only for demonstration purpose, not used for grading
# function is actually what is being tested
print(vfan('!', 5, '%'))

!        %
!!      %%
!!!    %%%
!!!!  %%%%
!!!!!%%%%%


In [11]:
# Sample 1.2
print(vfan('W', 11, 'M'))

W                    M
WW                  MM
WWW                MMM
WWWW              MMMM
WWWWW            MMMMM
WWWWWW          MMMMMM
WWWWWWW        MMMMMMM
WWWWWWWW      MMMMMMMM
WWWWWWWWW    MMMMMMMMM
WWWWWWWWWW  MMMMMMMMMM
WWWWWWWWWWWMMMMMMMMMMM


<h2 style='font-family: Verdana'>Question 2 - Logging with Lists</h2>

Write a function `every_n(ls, n)`, which takes a non-empty list `ls` and a positive integer `n`, and return the elements whose log base `n` is an integer.

For example, `every_n([1,2,4,6,8,10], 2)` will return `[1, 2, 4, 8]`.

The function code should take no more than one line without the use of semicolons.

You can assume that every element in ls is numeric and is positive (i.e., greater than zero) and `n` is greater than 1.

Note: If you are unable to solve using one line or without use of additional libraries, you may answer using multiple lines or importing libraries; 50% of the marks will still be awarded.

In [12]:
# you are allowed to use the math library for this question and this question only
import math

def every_n(ls, n) -> list:
    return [x for x in ls if math.log(x, n).is_integer()]


In [13]:
# Sample 2.1
every_n([1,2,4,6,8,10], 2)

[1, 2, 4, 8]

In [14]:
# Sample 2.2
every_n([.1,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,21,24,27,30], 3)

[3, 9, 27]

<h2 style='font-family: Verdana'>Question 3 - Sorting dictionaries</h2>

Write a function `sorted_dic` that takes in a dictionary whose keys are items (string) and values are currency values (string), and returns a dictionary with the same key-value pairs except that the values are sorted in descending order.

You can assume each currency value starts with `$` and has precision up to six decimal places.

The function code should take no more than one line without the use of semicolons.

Note: If you are unable to solve using one line or without use of additional libraries, you may answer using multiple lines or importing libraries; 50% of the marks will still be awarded.

In [15]:
def sorted_dic(menu) -> dict:
    return {k: v for k, v in sorted(menu.items(), key=lambda item: float(item[1][1:]), reverse=True)}


In [16]:
# Sample 3.1
menu1 = {
    'grilled fish': '$12.50',
    'fish & chips': '$9.50',
    'smoked salmon': '$14.70',
    'raw fish': '$2.99',
    'poached cod': '$20.10',
    'seared pollock': '$16.80'
}
sorted_dic(menu1)

{'poached cod': '$20.10',
 'seared pollock': '$16.80',
 'smoked salmon': '$14.70',
 'grilled fish': '$12.50',
 'fish & chips': '$9.50',
 'raw fish': '$2.99'}

In [17]:
# Sample 3.2
menu2 = {
    'fish ball': '$1.20',
    'prawn ball': '$1.70',
    'lobster ball': '$2.40',
    'crab ball': '$2.20',
    'cuttlefish ball': '$1.30',
}
sorted_dic(menu2)

{'lobster ball': '$2.40',
 'crab ball': '$2.20',
 'prawn ball': '$1.70',
 'cuttlefish ball': '$1.30',
 'fish ball': '$1.20'}

<h2 style='font-family: Verdana'>Question 4 - Sum the bounce</h2>

On a hypothetical planet, a ball is released from a height $h$ will drop down, hit the ground and then bounce back up to a height $h \times alpha$, where $alpha$ is a value between 0 and 1 exclusive, i.e., (0, 1).

Of course, once it has bounced up, the ball will fall back down and then bounce up again.

For example, if $h = 100$ and $alpha = 0.5$, then the ball would have travelled $100 + 50 + 50 + 25 + 25 + 12.5 + 12.5$ when it hits the ground the fourth time.

Write a function `sum_bounce(h, alpha, n)` where `h` is the starting height (positive, greater than zero), `alpha` as explained and `n` is the number of times the ball hits the ground. The return value is the total distance travelled.

You can assume `n` to be a positive integer (greater than zero).

In [24]:
def sum_bounce(h, alpha, n) -> float:
    distance = h
    for _ in range(n):
        distance += 2 * alpha * h
        h *= alpha
    return distance


In [25]:
# Sample 4.1
sum_bounce(100, .5, 4)

287.5

In [26]:
# Sample 4.2
sum_bounce(20, .2, 100)

29.999999999999996

<h2 style='font-family: Verdana'>Question 5 - Palindrome Clock Time</h2>

A digital clock which displays time in the format of hh:mm:ss (hour-hour in 24 hours format, minute-minute, and second-second) will occasionally show a string that looks the same backward as forward, which we call a palindromic clock time. For example, 12:33:21. In this question, your task is to list all palindromic clock time in chronological order for a given interval defined by two times.

Write a function named `find_palindrome_times` which takes in `start_time` and `end_time` as string arguments in the format of hh:mm, and returns a list of palindromic clock time between the two times inclusive, or an empty list if none found.

You can assume that the times will be valid times (00:00 to 23:59), and `start_time` will always be earlier than the `end_time`.

Note: If you are unable to solve without use of additional libraries, you may answer by importing libraries; 50% of the marks will still be awarded.

In [27]:
def find_palindrome_times(start_time, end_time) -> list:
    palindromes = []
    hour, minute = map(int, start_time.split(':'))
    
    while True:
        current_time = f'{hour:02d}:{minute:02d}'
        if current_time == current_time[::-1]:
            palindromes.append(current_time)
        
        if hour == int(end_time[:2]) and minute == int(end_time[3:]):
            break
        
        minute += 1
        if minute == 60:
            minute = 0
            hour += 1
            if hour == 24:
                hour = 0
    
    return palindromes


In [28]:
# Sample 5.1
find_palindrome_times('00:01', '00:59')

[]

In [29]:
# Sample 5.2
find_palindrome_times('23:00', '23:59')

['23:32']

<h2 style='font-family: Verdana'>Question 6 - Fraction to Decimal</h2>

A mixed number is a whole number and a proper fraction represented together. It generally represents a number between any two whole numbers. Examples of mixed numbers are as following:

$$
1 \frac{1}{2}, -7\frac{3}{4}, 20\frac{5}{7}
$$

In the one of the above examples, $1$ $1/2$, 1 is the whole number, and $\frac{1}{2}$ is the fraction.

Write a function `frac_to_dec` that takes in a string input of a mixed number and return the decimal value. For example, $1$ $1/2$ = 1.5.

The function code should not be longer than five lines without the use of semicolons.

You may assume that the mixed number can have an *optional* whole number (can be positive or negative or zero) and the numerator and denominator of the fraction are positive integers.
You may also assume that there are no invalid characters.

Note: If you are unable to solve using at most five lines or without use of additional libraries, you may answer using more lines or importing libraries; 50% of the marks will still be awarded.

In [39]:
def frac_to_dec(frac) -> float:
    parts = frac.split()
    numerator, denominator = map(int, parts[-1].split('/'))
    return round(sum(map(int, parts[:-1])) + numerator / denominator, 5)


In [40]:
# Sample 6.1
frac_to_dec('1 1/2')

1.5

In [41]:
# Sample 6.2
frac_to_dec('1/99')

0.0101

<h2 style='font-family: Verdana'>Question 7 - Sinful Treats</h2>

Study the `NewYearGoodie` class. Each instance of `NewYearGoodie` has a name, calories per gram and weight in gram. You can assume calories and weight are positive (i.e., more than zero) quantities.

Create a function called `isMoreSinful` that takes a varying number of treat instances and returns the name of treat with the highest caloric content *in total*.
If multiple treats have the same caloric content, return the **first** treat according to the inputs of the function.

You can assume the function will receive at least one input instance of NewYearGoodie

In [48]:
# do not change anything in this cell
class NewYearGoodie:
    def __init__(self, name, calories, weight):
        self.name = name
        self.calories = calories
        self.weight = weight

In [49]:
### Your code here ###
# the name of the function is isMoreSinful
def isMoreSinful(*treats) -> str:
    treat_calories = {}

    for treat in treats:
        if treat.name not in treat_calories:
            treat_calories[treat.name] = treat.calories * treat.weight
        else:
            treat_calories[treat.name] += treat.calories * treat.weight

    most_sinful_treat = max(treat_calories, key=treat_calories.get)
    return most_sinful_treat


In [50]:
# Sample 7.1
isMoreSinful(NewYearGoodie('Pineapple Tarts', 82.5, 20),
            NewYearGoodie('Bak Kwa', 115, 28),
            NewYearGoodie('Pineapple Tarts', 82.5, 20))

'Pineapple Tarts'

In [51]:
# Sample 7.2
isMoreSinful(NewYearGoodie('Pineapple Tarts', 82.5, 20),
            NewYearGoodie('Love Letters', 56.5, 13),
            NewYearGoodie('Pure Candy Cube', 165, 10.0))

'Pineapple Tarts'

<h2 style='font-family: Verdana'>Question 8 - Inheriting the Rational</h2>

Write a class called `Rational` which represents a rational number (i.e., a fraction). You do not have to worry about repesenting the fraction in proper form.

The constructor (`__init__`) takes in two integers, one being the numerator, the other being the denominator. You can assume the denominator to be non-zero.

The string (`__str__`) presents the rational number in string format, and code has been provided.

The addition (and by extension, subtraction) operations of rational is performed as follows:

$$
\frac{a}{b} + \frac{c}{d} = \frac{ad + bc}{bd}
$$

Your `Rational` class should implement an `add` method to perform the addition operation and return the resulting `Rational` instance.

Then create a child class called `Integer` which represents an integer number.
The constructor of the `Integer` class takes in only one integer value.

The `Integer` class automatically inherits the addition method of the `Rational` class.

In [52]:
class Rational:
    def __init__(self, n, d):
        self.n = n
        self.d = d
    
    def add(self, another_r):
        new_n = self.n * another_r.d + self.d * another_r.n
        new_d = self.d * another_r.d
        return Rational(new_n, new_d)

    def __str__(self):
        return f'{self.n} / {self.d}'

class Integer(Rational):
    def __init__(self, n):
        super().__init__(n, 1)
        
def __str__(self):
        return str(self.n)

The test cases will still be examining the numerical contents of your object instances, but printing the objects makes it easy to verify that your operations work as intended.

In [53]:
# Sample 8.1
print(Integer(1).add(Integer(2)))

3 / 1


In [54]:
# Sample 8.2
print(Rational(1,2).add(Integer(-2)))

-3 / 2
